staging/src/k8s.io/api/resource/v1alpha3/types.go

```diff
@@ -0,0 +1,998 @@
+type ResourceSlice struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard object metadata
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// Contains the information published by the driver.
+	//
+	// Changing the spec bumps up the generation number.
+	Spec ResourceSliceSpec `json:"spec" protobuf:"bytes,2,name=spec"`
+
+	// Future extension: status.
+}
+
+// ResourceSliceSpec contains the information published by the driver in one ResourceSlice.
+type ResourceSliceSpec struct {
+	// DriverName identifies the DRA driver providing the capacity information.
+	// A field selector can be used to list only ResourceSlice
+	// objects with a certain driver name.
+	//
+	// Must be a DNS subdomain and should end with a DNS domain owned by the
+	// vendor of the driver.
+	//
+	// +required
+	DriverName string `json:"driverName" protobuf:"bytes,1,name=driverName"`
+
+	// PoolName is used to identify devices. For node-local devices, this
+	// is often the node name, but this is not required.
+	//
+	// It must not be longer than 253 and must consist of one or more DNS sub-domains
+	// separated by slashes.
+	//
+	// +required
+	PoolName string `json:"poolName" protobuf:"bytes,2,name=poolName"`
+
+	// NodeName identifies the node which provides the devices.
+	// A field selector can be used to list only ResourceSlice
+	// objects belonging to a certain node.
+	//
+	// This field can be used to limit access from nodes to slices with
+	// the same node name. It also indicates to autoscalers that adding
+	// new nodes of the same type as some old node might also make new
+	// devices available.
+	//
+	// NodeName and NodeSelector are mutually exclusive. One of them
+	// must be set.
+	//
+	// +optional
+	NodeName *string `json:"nodeName,omitempty" protobuf:"bytes,3,opt,name=nodeName"`
+
+	// Defines which nodes have access to the devices in the pool.
+	// If the node selector is empty, all nodes have access.
+	//
+	// NodeName and NodeSelector are mutually exclusive. One of them
+	// must be set.
+	//
+	// +optional
+	NodeSelector *v1.NodeSelector `json:"nodeSelector,omitempty" protobuf:"bytes,4,opt,name=nodeSelector"`
+
+	// The generation gets bumped in all slices of a pool whenever device
+	// definitions change. A consumer must only use device definitions from slices
+	// with the highest generation number and ignore all others.
+	PoolGeneration int64 `json:"poolGeneration" protobuf:"bytes,5,name=poolGeneration"`
+
+	// The total number of slices in the pool.
+	// Consumers can use this to check whether they have
+	// seen all slices.
+	PoolSliceCount int64 `json:"poolSliceCount" protobuf:"bytes,6,name=poolSliceCount"`
+
+	// Devices lists all available devices in this pool.
+	//
+	// Must not have more than 128 entries.
+	//
+	// +required
+	// +listType=atomic
+	Devices []Device `json:"devices" protobuf:"bytes,7,name=devices"`
+
+	// FUTURE EXTENSION: some other kind of list, should we ever need it.
```

Comment:
Do you think these future extensions are one-of with Devices, or that a single slice might hold both devices and gizmos?

Reply:
I was actually thinking of a second slice with a different type.

Then a consumer which doesn't know about that new slice field will ignore it, which is what we want: it can allocate what it knows about, without doing something wrong about something that it does not fully understand.

In https://github.com/kubernetes-sigs/wg-device-management/pull/36/files, @johnbelamaric is proposing to add a `SliceContent string` enum.

We could add that to cover the scenario that we want to extend the `Device` struct in the future. We need a way to warn a consumer based on 1.31 that a 1.32 ResourceSlice has new unknown fields and that it better shouldn't work with the slice. That is how we can add the partitionable devices as an extension of the `Device` struct in a safe way instead of having to add a new `PartitionableDevices []PartitionableDevice` slice.

I'm not fully convinced that it is needed, but if we want it, we have to add it now.

Reply:
> If we force it to be two slices, then that controller will see a slice with empty Devices[] which is strictly allowed, right?

Yes. That is what I would do instead of adding fields to the `Device` struct.

IMHO the discriminator is only needed if we want to add fields to that struct *and* don't want old clients to ignore them, which would be the case for partitionable devices.

> FWIW in this PR, I would just drop the comments about future extension and clarify that the Devices[] list is allowed to be empty (optional)

Ack.

Reply:
So let's use this to discuss how the current API gets extended to support partitionable devices. I see two options:

- Add new fields to the `Device` struct in `Devices []Device`. In this case we need a discriminator because the current scheduler would ignore the new `Device` fields, which woudn't be okay.
- Add a `PartitionableDevice` struct that has the same fields as `Device` or (perhaps better, from an implementation perspective) embeds that struct, plus the additional fields. No discriminator needed, current scheduler cannot allocate the partitionable devices because it doesn't even see them.

In both cases the `ResourceSlice` itself needs new fields that define what is being shared between partitions.

The advantage of the first option is that we need less structs, but I don't like the discriminator. What if we add even more fields? At some point the value of that discriminator will be "DevicesWithPartitionsButMaybeAlsoSomethingElse"...

The second option needs an additional struct which is a bit more work when initializing in Go, but I think that is okay. I prefer this option - but no strong opinion.

Reply:
The discussion moved to https://github.com/kubernetes/kubernetes/pull/125488#discussion_r1668553032.